File: gallery/pygallery/loader.py
"""Building an app's Shiny app the first time somebody asks for its URL.

Fifteen apps importing numpy, and reading their data, at startup is how a
deployment walks into the Connect Cloud startup timeout. Nothing here runs
until a request arrives for that app's mount, and then it runs once.
"""

from __future__ import annotations

import importlib.util
import sys
from pathlib import Path
from types import ModuleType
from typing import TYPE_CHECKING

import anyio
import anyio.to_thread
from htmltools import tags
from shiny import App
from shinyreact import ReactApp, page_react
from starlette.types import Receive, Scope, Send

if TYPE_CHECKING:
    from .registry import Entry


class LazyApp:
    """An ASGI app that becomes `apps/<slug>` on first use."""

    def __init__(self, entry: Entry) -> None:
        self._entry = entry
        self._app: App | None = None
        self._lock = anyio.Lock()

    async def __call__(self, scope: Scope, receive: Receive, send: Send) -> None:
        if self._app is None:
            async with self._lock:
                # Checked again inside the lock: two requests can both find
                # None before either takes it.
                if self._app is None:
                    # In a worker thread. Importing numpy and reading a data
                    # file is seconds of blocking CPU, and on the event loop
                    # that stalls every other app's websocket, including the
                    # gallery page itself.
                    self._app = await anyio.to_thread.run_sync(self._build)

        await self._app(scope, receive, send)

    def _build(self) -> App:
        directory = self._entry.directory

        # What `shiny run` does through uvicorn's app_dir, so that
        # `from cell_atlas import ...` inside app.py resolves exactly as it
        # does when the app runs on its own.
        sys.path.insert(0, str(directory))
        try:
            module = _import_module(
                f"showcase_{self._entry.module_id}", directory / "app.py"
            )
        finally:
            sys.path.remove(str(directory))

        server = getattr(module, "server", None)
        if not callable(server):
            raise RuntimeError(
                f"apps/{self._entry.slug}/app.py defines no `server` function. "
                "The gallery composes `server` with a page of its own. See "
                "CONTRIBUTING.md, 'The contract every app follows'."
            )

        # Built here rather than reusing module.app, so the gallery can add
        # the back link and pin src_dir to an absolute path. The app's own
        # ReactApp() call is for running it standalone.
        return ReactApp(
            server,
            ui=lambda _request: page_react(
                _back_link(),
                _back_link_style(),
                src_dir=directory / "www",
                title=self._entry.title,
            ),
        )


def _import_module(name: str, path: Path) -> ModuleType:
    spec = importlib.util.spec_from_file_location(name, path)
    if spec is None or spec.loader is None:
        raise ImportError(f"Cannot load {path}")
    module = importlib.util.module_from_spec(spec)
    # Registered before execution, because dataclasses and pickle look
    # themselves up by module name while the module is still running.
    sys.modules[name] = module
    spec.loader.exec_module(module)
    return module


def _back_link():
    # Relative, because the app is mounted at /app/<slug>/ and the gallery is
    # two levels up. An absolute /  would break behind a proxy that serves
    # the content under a path of its own.
    return tags.a(
        "Back to the gallery",
        href="../../",
        class_="gallery-back",
    )


def _back_link_style():
    # Inline rather than in fifteen client bundles. The same www/ui.js serves
    # the app standalone, where there is no gallery to go back to, so this
    # belongs to the gallery and not to the app.
    return tags.style("""
      .gallery-back {
        position: fixed;
        left: 12px;
        bottom: 12px;
        z-index: 1000;
        padding: 6px 12px;
        border-radius: 999px;
        border: 1px solid rgba(0, 0, 0, 0.12);
        background: rgba(255, 255, 255, 0.92);
        color: #16181d;
        font: 13px system-ui, -apple-system, "Segoe UI", sans-serif;
        text-decoration: none;
        box-shadow: 0 1px 4px rgba(0, 0, 0, 0.12);
      }
      .gallery-back:hover {
        border-color: rgba(0, 0, 0, 0.28);
      }
    """)

File: gallery/R/loader.R
# Loading one app into the gallery process.
#
# Greenfield apps here are written to be loaded, so there is nothing to shim.
# Every app defines `ui` and `server` at top level, reads its own files
# through APP_DIR, and ends with shinyApp(ui, server) only so that runApp()
# works on it directly. The gallery supplies APP_DIR and reads the two
# objects back out.
#
# An app is loaded at most once per process, on the first visit to it.
#
# R is single threaded, so that first load blocks every other session while
# it runs. Keep top level work in an app.R under a second. The Python gallery
# does the same import in a worker thread and blocks nobody, which is a real
# difference between the two deployments and not something to paper over.

.loaded <- new.env(parent = emptyenv())

load_app <- function(slug) {
  hit <- .loaded[[slug]]
  if (!is.null(hit)) {
    return(hit)
  }

  entry <- REGISTRY$entries[[slug]]
  if (is.null(entry)) {
    stop("No app called ", slug, " in the gallery registry.")
  }

  # Parent is globalenv() so the app sees every attached package, while its
  # own objects stay out of every other app's way.
  env <- new.env(parent = globalenv())
  env$APP_DIR <- entry$directory

  # What runApp() auto-loads for an app running on its own.
  shiny::loadSupport(entry$directory, renv = env, globalrenv = env)

  # Belt and braces for any top level relative read the contract did not
  # anticipate. R is single threaded and Shiny runs one session's code at a
  # time, so no other session sees this working directory.
  old_wd <- setwd(entry$directory)
  on.exit(setwd(old_wd), add = TRUE)

  sys.source(
    file.path(entry$directory, "app.R"),
    envir = env,
    keep.source = FALSE
  )

  if (is.null(env$ui) || !is.function(env$server)) {
    stop(
      "apps/",
      slug,
      "/app.R must define `ui` and `server` at top level. ",
      "The gallery composes them with a page of its own. See CONTRIBUTING.md, ",
      "'The contract every app follows'."
    )
  }

  loaded <- list(slug = slug, ui = env$ui, server = env$server, env = env)
  .loaded[[slug]] <- loaded
  loaded
}

# The counter, on every page the gallery serves.
#
# It goes here rather than into the app bundles for the same reason the back
# link does: the same www/ui.js also serves the app standalone, where there is
# nothing to count. Blank turns it off.
#
# count.js drops requests from local addresses on its own, so running this on
# 127.0.0.1 never reaches the stats.
gallery_analytics <- function(url = REGISTRY$site$analytics %||% "") {
  if (is.null(url) || !nzchar(url)) {
    return(NULL)
  }
  htmltools::tags$script(
    `data-goatcounter` = url,
    async = NA,
    src = "//gc.zgo.at/count.js"
  )
}

# The link back to the gallery, added to every app page.
#
# It lives here rather than in fifteen client bundles, because the same
# www/ui.js also serves the app on its own, where there is no gallery to go
# back to.
gallery_back_link <- function() {
  # Top right, because that is where a reader looks for a way out and because
  # every app in this repo lays its own content out left aligned under a max
  # width, so the top right corner is the one place a floating control is
  # never in the way. Bottom left, where this used to be, was missed entirely.
  htmltools::tagList(
    htmltools::tags$a(
      htmltools::tags$span(
        htmltools::HTML("&larr;"),
        class = "arrow",
        `aria-hidden` = "true"
      ),
      "Back to the gallery",
      href = "?",
      class = "gallery-back"
    ),
    htmltools::tags$style(htmltools::HTML(
      ".gallery-back {
         position: fixed;
         right: 16px;
         top: 14px;
         z-index: 2000;
         display: inline-flex;
         align-items: center;
         gap: 6px;
         padding: 7px 14px 7px 11px;
         border-radius: 999px;
         border: 1px solid rgba(15, 23, 42, 0.12);
         background: #ffffff;
         color: #16181d;
         font: 500 13px system-ui, -apple-system, 'Segoe UI', sans-serif;
         text-decoration: none;
         box-shadow: 0 2px 10px rgba(16, 24, 40, 0.14);
         transition: box-shadow 120ms ease, border-color 120ms ease;
       }
       .gallery-back:hover {
         border-color: rgba(47, 111, 237, 0.55);
         box-shadow: 0 4px 16px rgba(16, 24, 40, 0.2);
       }
       .gallery-back:focus-visible {
         outline: 2px solid #2f6fed;
         outline-offset: 2px;
       }
       .gallery-back .arrow { color: #2f6fed; font-size: 14px; line-height: 1; }
       @media (max-width: 900px) {
         .gallery-back {
           top: 8px;
           right: 8px;
           padding: 5px 11px 5px 9px;
           font-size: 12px;
         }
       }"
    ))
  )
}
